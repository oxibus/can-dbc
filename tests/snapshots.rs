#![cfg(all(feature = "serde", feature = "encodings"))]

use nom::Err;
use std::borrow::Cow;
use std::fs;
use std::path::{Path, PathBuf};

use can_dbc::{decode_cp1252, Dbc, Error};
use insta::{assert_debug_snapshot, assert_yaml_snapshot, with_settings};
use test_each_file::test_each_path;

struct TestConfig {
    test_root: &'static str,
    snapshot_suffix: &'static str,
    use_cp1251: bool,
    create_snapshot: bool,
}

static TEST_DIRS: &[TestConfig] = &[
    TestConfig {
        test_root: "shared-test-files",
        snapshot_suffix: "",
        use_cp1251: true,
        create_snapshot: true,
    },
    TestConfig {
        test_root: "opendbc/opendbc/dbc",
        snapshot_suffix: "opendbc",
        use_cp1251: false,
        create_snapshot: false,
    },
];

test_each_path! { for ["dbc"] in "./tests/fixtures/opendbc/opendbc/dbc" as dbc => parse_one_file }
test_each_path! { for ["dbc"] in "./tests/fixtures/shared-test-files" as shared => parse_one_file }
// upper case extension
test_each_path! { for ["DBC"] in "./tests/fixtures/shared-test-files" as shared2 => parse_one_file }

/// Get `test root`, `snapshot name suffix`, `use cp1251`, `create snapshot` for the given path
fn get_test_info(path: &Path) -> Option<(PathBuf, &'static TestConfig)> {
    if !path
        .extension()
        .unwrap_or_default()
        .eq_ignore_ascii_case("dbc")
    {
        return None;
    }
    let path_str = path.to_str().unwrap();
    let parent = path.parent().unwrap();
    for item in TEST_DIRS {
        // Ensure slashes are there for easier matching
        let test_root = format!("/{}/", item.test_root);
        let mut path_dir = parent.to_str().unwrap().to_string();
        if !path_dir.ends_with('/') {
            path_dir.push('/');
        }
        if let Some(pos) = path_dir.find(&test_root) {
            let parent = PathBuf::from("snapshots")
                .join(item.snapshot_suffix)
                .join(&path_dir[pos + test_root.len()..]);
            return Some((parent, item));
        }
    }
    panic!("Unknown test directory: {path_str}");
}

/// Test parsing all DBC files
#[test]
fn test_if_submodules_are_present() {
    for test in TEST_DIRS {
        let dir = Path::new("./tests/fixtures").join(test.test_root);
        fs::read_dir(&dir).unwrap_or_else(|e| {
            let dir_display = dir.display();
            panic!(
                "
--------------------------------------------------------------------------
Error reading dbc test files from   {dir_display}
{e}
Make sure git submodules are up to date by running
    git submodule update --init --recursive
--------------------------------------------------------------------------
"
            )
        });
    }
}

/// Parse a single DBC file and assert a snapshot of the result.
fn parse_one_file([path]: [&Path; 1]) {
    let Some((
        snapshot_path,
        &TestConfig {
            use_cp1251,
            create_snapshot,
            ..
        },
    )) = get_test_info(path)
    else {
        return;
    };

    let file_name = path.file_stem().unwrap().to_string_lossy().to_string();
    let buffer = fs::read(path).unwrap();
    let buffer = if use_cp1251 {
        decode_cp1252(&buffer)
            .unwrap_or_else(|| panic!("Failed to decode {} as cp1252", path.display()))
    } else {
        Cow::Borrowed(
            std::str::from_utf8(&buffer)
                .unwrap_or_else(|_| panic!("Failed to decode {} as utf-8", path.display())),
        )
    };

    report_test_result(
        Dbc::try_from(buffer.as_ref()),
        snapshot_path,
        file_name,
        create_snapshot,
    );
}

fn report_test_result(
    result: Result<Dbc, Error>,
    snapshot_path: PathBuf,
    file_name: String,
    do_snapshot: bool,
) {
    let do_snapshot = std::env::var("SKIP_INSTA").is_err()
        && (do_snapshot || std::env::var("FORCE_INSTA").is_ok());

    with_settings! {
        {
            omit_expression => true,
            snapshot_path => snapshot_path,
            prepend_module_to_snapshot => false
        },
        {
            match result {
                Ok(result) => {
                    if do_snapshot {
                        assert_yaml_snapshot!(file_name, result);
                    } // else ignore successful results when not snapshotting
                }
                Err(e) => {
                    let file_name = format!("!error___{file_name}");
                    assert_debug_snapshot!(file_name, err_to_str(&e));
                }
            }
        }
    }
}

fn err_to_str<'a>(err: &'a Error) -> &'a str {
    match err {
        Error::Incomplete(_, _) => "incomplete",
        Error::Nom(e) => match e {
            Err::Incomplete(_) => "nom-incomplete",
            Err::Error(_) => "nom-error",
            Err::Failure(_) => "nom-failure",
        },
        Error::MultipleMultiplexors => "multiple multiplexors",
    }
}
