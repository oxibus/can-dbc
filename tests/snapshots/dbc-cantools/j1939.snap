---
source: tests/snapshots.rs
---
version: ""
new_symbols:
  - NS_DESC_
  - CM_
  - BA_DEF_
  - BA_
  - VAL_
  - CAT_DEF_
  - CAT_
  - FILTER
  - BA_DEF_DEF_
  - EV_DATA_
  - ENVVAR_DATA_
  - SGTYPE_
  - SGTYPE_VAL_
  - BA_DEF_SGTYPE_
  - BA_SGTYPE_
  - SIG_TYPE_REF_
  - VAL_TABLE_
  - SIG_GROUP_
  - SIG_VALTYPE_
  - SIGTYPE_VALTYPE_
  - BO_TX_BU_
  - BA_DEF_REL_
  - BA_REL_
  - BA_DEF_DEF_REL_
  - BU_SG_REL_
  - BU_EV_REL_
  - BU_BO_REL_
  - SG_MUL_VAL_
bit_timing: []
nodes:
  - Node2
  - Node1
messages:
  - id:
      Extended: 355729921
    name: Message1
    size: 8
    transmitter:
      NodeName: Node1
    signals:
      - name: Signal1
        multiplexer_indicator: Plain
        start_bit: 0
        size: 8
        byte_order: LittleEndian
        value_type: Signed
        factor: 1
        offset: 0
        min: 0
        max: 0
        unit: ""
        receivers:
          - Node2
  - id:
      Extended: 368054274
    name: Message2
    size: 8
    transmitter:
      NodeName: Node2
    signals:
      - name: Signal2
        multiplexer_indicator: Plain
        start_bit: 0
        size: 8
        byte_order: LittleEndian
        value_type: Signed
        factor: 1
        offset: 0
        min: 0
        max: 0
        unit: ""
        receivers:
          - Node1
attribute_definitions:
  - Message:
      - TpJ1939VarDlc
      - Enum:
          - "No"
          - "Yes"
  - Signal:
      - SigType
      - Enum:
          - Default
          - Range
          - RangeSigned
          - ASCII
          - Discrete
          - Control
          - ReferencePGN
          - DTC
          - StringDelimiter
          - StringLength
          - StringLengthControl
          - MessageCounter
          - MessageChecksum
  - Signal:
      - GenSigEVName
      - String
  - Signal:
      - GenSigILSupport
      - Enum:
          - "No"
          - "Yes"
  - Signal:
      - GenSigSendType
      - Enum:
          - Cyclic
          - OnWrite
          - OnWriteWithRepetition
          - OnChange
          - OnChangeWithRepetition
          - IfActive
          - IfActiveWithRepetition
          - NoSigSendType
  - Message:
      - GenMsgFastOnStart
      - Int:
          - 0
          - 100000
  - Signal:
      - GenSigInactiveValue
      - Int:
          - 0
          - 0
  - Message:
      - GenMsgCycleTimeFast
      - Int:
          - 0
          - 3600000
  - Message:
      - GenMsgNrOfRepetition
      - Int:
          - 0
          - 1000000
  - Signal:
      - GenSigStartValue
      - Int:
          - 0
          - 2147483647
  - Message:
      - GenMsgDelayTime
      - Int:
          - 0
          - 1000
  - Message:
      - GenMsgILSupport
      - Enum:
          - "No"
          - "Yes"
  - Message:
      - GenMsgStartDelayTime
      - Int:
          - 0
          - 100000
  - Node:
      - NodeLayerModules
      - String
  - Node:
      - ECU
      - String
  - Node:
      - NmJ1939SystemInstance
      - Int:
          - 0
          - 15
  - Node:
      - NmJ1939System
      - Int:
          - 0
          - 127
  - Node:
      - NmJ1939ManufacturerCode
      - Int:
          - 0
          - 2047
  - Node:
      - NmJ1939IndustryGroup
      - Int:
          - 0
          - 7
  - Node:
      - NmJ1939IdentityNumber
      - Int:
          - 0
          - 2097151
  - Node:
      - NmJ1939FunctionInstance
      - Int:
          - 0
          - 7
  - Node:
      - NmJ1939Function
      - Int:
          - 0
          - 255
  - Node:
      - NmJ1939ECUInstance
      - Int:
          - 0
          - 3
  - Node:
      - NmJ1939AAC
      - Int:
          - 0
          - 1
  - Node:
      - NmStationAddress
      - Int:
          - 0
          - 255
  - Message:
      - GenMsgSendType
      - Enum:
          - cyclic
          - NotUsed
          - IfActive
          - NotUsed
          - NotUsed
          - NotUsed
          - NotUsed
          - NotUsed
          - noMsgSendType
  - Message:
      - GenMsgRequestable
      - Int:
          - 0
          - 1
  - Message:
      - GenMsgCycleTime
      - Int:
          - 0
          - 3600000
  - Signal:
      - SPN
      - Int:
          - 0
          - 524287
  - Plain:
      - DBName
      - String
  - Plain:
      - BusType
      - String
  - Plain:
      - ProtocolType
      - String
  - Message:
      - VFrameFormat
      - Enum:
          - StandardCAN
          - ExtendedCAN
          - reserved
          - J1939PG
attribute_defaults:
  - name: TpJ1939VarDlc
    value:
      String: "No"
  - name: SigType
    value:
      String: Default
  - name: GenSigEVName
    value:
      String: Env@Nodename_@Signame
  - name: GenSigILSupport
    value:
      String: "Yes"
  - name: GenSigSendType
    value:
      String: NoSigSendType
  - name: GenMsgFastOnStart
    value:
      Double: 0
  - name: GenSigInactiveValue
    value:
      Double: 0
  - name: GenMsgCycleTimeFast
    value:
      Double: 0
  - name: GenMsgNrOfRepetition
    value:
      Double: 0
  - name: GenSigStartValue
    value:
      Double: 0
  - name: GenMsgDelayTime
    value:
      Double: 0
  - name: GenMsgILSupport
    value:
      String: "Yes"
  - name: GenMsgStartDelayTime
    value:
      Double: 0
  - name: NodeLayerModules
    value:
      String: ""
  - name: ECU
    value:
      String: ""
  - name: NmJ1939SystemInstance
    value:
      Double: 0
  - name: NmJ1939System
    value:
      Double: 0
  - name: NmJ1939ManufacturerCode
    value:
      Double: 0
  - name: NmJ1939IndustryGroup
    value:
      Double: 0
  - name: NmJ1939IdentityNumber
    value:
      Double: 0
  - name: NmJ1939FunctionInstance
    value:
      Double: 0
  - name: NmJ1939Function
    value:
      Double: 0
  - name: NmJ1939ECUInstance
    value:
      Double: 0
  - name: NmJ1939AAC
    value:
      Double: 0
  - name: NmStationAddress
    value:
      Double: 254
  - name: GenMsgSendType
    value:
      String: noMsgSendType
  - name: GenMsgRequestable
    value:
      Double: 1
  - name: GenMsgCycleTime
    value:
      Double: 0
  - name: SPN
    value:
      Double: 0
  - name: DBName
    value:
      String: ""
  - name: BusType
    value:
      String: CAN
  - name: ProtocolType
    value:
      String: J1939
  - name: VFrameFormat
    value:
      String: J1939PG
attribute_values:
  - name: DBName
    value:
      Raw:
        String: foo
  - name: NmStationAddress
    value:
      NetworkNode:
        - Node2
        - Double: 2
  - name: NmStationAddress
    value:
      NetworkNode:
        - Node1
        - Double: 1
  - name: VFrameFormat
    value:
      MessageDefinition:
        - Extended: 355729921
        - Double: 3
  - name: SPN
    value:
      Signal:
        - Extended: 355729921
        - Signal1
        - Double: 500
